### PR TITLE
Add support for per-source and total external condition caps.

### LINKED ISSUE #12720: Stackable external condition duration cannot be reset by the source, a new instance will be applied instead.
This is a regression due to #8658 became undone during the upgrades->conditions rewrite.

### PR DESCRIPTION
**Pull request by pchote**
This removes the concept of an "external condition" from `ConditionManager`, instead having the `ExternalCondition` trait act as an intermediary.  `ExternalCondition` keeps track of the different external sources, and can be configured with caps on the total and per-source condition levels that are granted.

Fixes #12720.

### CODE REVIEW

**Comment by pchote** [comment]
For a testcase, change `^EmpDisable` in TS to:
```yaml
^EmpDisable:
	StackedCondition@emp:
		Condition: empdisable
		StackedConditions: empdisable2
	WithColoredOverlay@EMPDISABLE:
		RequiresCondition: empdisable && !empdisable2
		Palette: highlight
	WithColoredOverlay@EMPDISABLE2:
		RequiresCondition: empdisable2
		Palette: disabled
	DisableOnCondition@EMPDISABLE:
		RequiresCondition: empdisable
	TimedConditionBar@EMPDISABLE:
		Condition: empdisable
		Color: FFFFFF
	WithIdleOverlay@EMPDISABLE:
		Sequence: emp-overlay
		Palette: effect
		RequiresCondition: empdisable
		ShowToEnemies: true
		ZOffset: 512
	PowerMultiplier@EMPDISABLE:
		RequiresCondition: empdisable
		Modifier: 0
	ExternalCondition@EMPDISABLE:
		Condition: empdisable
```

Build an EMP cannon and, with cheats enabed, observe it first turn light and then dark like my older stacked condition test cases.

Now add `SourceCap: 1` to `ExternalCondition`.  Notice that a single cannon can only activate the first level.  Build a second cannon and have that attack the target immediately after the first (sell the first one after it fires so the second one can fire).  Notice that the two cannons do indeed stack to level 2.

Now change `SourceCap: 1` to `TotalCap: 1`.   Notice that even with two cannons you can never advance past the first level.

File: OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		}
+
+		public bool CanGrantCondition(Actor self, object source, int duration = 0)
+		{
+			if (conditionManager == null || source == null)
+				return false;
+
+			// Timed tokens do not count towards the source cap: the condition with the shortest
+			// remaining duration can always be revoked to make room.
+			if (Info.SourceCap > 0 && permanentTokens.GetOrAdd(source).Count >= Info.SourceCap)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs]: Doesn't this also need a && duration == 0 check then? (same applies to L64).

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs]: If the cap has been reached with just the permanent tokens then neither timed nor permanent conditions are allowed.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs]: Ah, permanentTokens. Nevermind me then. The value of duration went unused however.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs]: > The value of duration went unused however.

Good point, i'll fix that.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs]: Fixed.

**Review by GraionDilach — APPROVED** [review]
Looks good to me and I'm more than happy to see this regression fixed in such an elegant way it can't create problems anymore :+1:

**Review by reaperrr — APPROVED** [review]
Looks good to me and the test case works as intended.

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8762f7a5ba24f2ac4f0c17ef1a5023b93b18e95a...87cd837931ad41f1481800ce65627f0448965123)

**Pull request merged**